---
title: "07_ancestry"
author: "John Dou"
date: "January 20, 2021"
output: html_document
---

```{batch}
module load Bioinformatics
module load plink/1.9
module load R

cd /nfs/turbo/bakulski1/People/johndou/ALS_Goutman/07_Ancestry/
```


## merge with 1000 genomes

```{batch}
# Extract the variants present in als dataset from the 1000 genomes dataset.

awk '{print$2}' ../06_Relate/als10_PAR_rel.bim > als_SNPs.txt
plink --bfile /nfs/turbo/bakulski1/People/johndou/ALS/07_ancestry/1000G_QC/1000G_maf --extract als_SNPs.txt --allow-extra-chr --make-bed --out 1000G_snp_match



# Extract the variants present in 1000 Genomes dataset from the als dataset.

awk '{print$2}' /nfs/turbo/bakulski1/People/johndou/ALS/07_ancestry/1000G_QC/1000G_maf.bim > 1000G_snps.txt
plink --bfile ../06_Relate/als10_PAR_rel --extract 1000G_snps.txt --recode --make-bed --out als_snp_match



# The datasets must have the same build. Change the build to 1000 Genomes data build.

awk '{print$2,$4}' als_snp_match.map > build.txt
plink --bfile 1000G_snp_match --update-map build.txt --make-bed --out 1000G_snp_match_build



# 1) set reference genome 
awk '{print$2,$5}' 1000G_snp_match_build.bim > 1000G_ref_list.txt
plink --bfile als_snp_match --reference-allele 1000G_ref_list.txt --make-bed --out als_ref_match



# 2) Resolve strand issues.
# Check for potential strand issues.
awk '{print$2,$5,$6}' 1000G_snp_match_build.bim > 1000G_tmp
awk '{print$2,$5,$6}' als_ref_match.bim > als_tmp
sort 1000G_tmp als_tmp |uniq -u > all_differences.txt

## Flip SNPs for resolving strand issues.
# Print SNP-identifier and remove duplicates.
awk '{print$1}' all_differences.txt | sort -u > flip_list.txt
plink --bfile als_ref_match --flip flip_list.txt --reference-allele 1000G_ref_list.txt --make-bed --out corrected_als

# Check for SNPs which are still problematic after they have been flipped.
awk '{print$2,$5,$6}' corrected_als.bim > corrected_als_tmp
sort 1000G_tmp corrected_als_tmp |uniq -u  > uncorresponding_SNPs.txt



# 3) Remove problematic SNPs from als data and 1000 Genomes.
awk '{print$1}' uncorresponding_SNPs.txt | sort -u > SNPs_for_exlusion.txt

# Remove the remaining problematic SNPs from both datasets.
plink --bfile corrected_als --exclude SNPs_for_exlusion.txt --make-bed --out als_ready_for_merge
plink --bfile 1000G_snp_match_build --exclude SNPs_for_exlusion.txt --make-bed --out 1000G_ready_for_merge

# Merge als data with 1000 Genomes Data.
plink --bfile als_ready_for_merge --bmerge 1000G_ready_for_merge.bed 1000G_ready_for_merge.bim 1000G_ready_for_merge.fam --allow-no-sex --make-bed --out als_1000G_merge



## Perform MDS 
# Using a set of pruned SNPs
plink --bfile als_1000G_merge --extract ../05_Heterozygosity/indepSNP.prune.in --genome --out ./MDS/merge
plink --bfile als_1000G_merge --read-genome ./MDS/merge.genome --cluster --mds-plot 10 --out ./MDS/merge_mds

```


```{r}
tg_pop <- read.table("/scratch/png_project_root/png_project1/shared_data/1000genomes/ancestrydat/1000gene_pops.txt")
colnames(tg_pop) <- c('id','sex','superpop','pop')
saveRDS(tg_pop, file='tg_pop.RDA')

library(openxlsx)
als_pop <- read.table("als_ready_for_merge.fam")
als_pop <- als_pop[,c(2,3)]
colnames(als_pop) <- c('id','superpop')
als_pop$superpop <- "" 

tg_pop_m <- tg_pop[,c(1,3)]
pop_m <- rbind(tg_pop_m,als_pop)

data <- read.table(file="MDS/merge_mds.mds",header=TRUE)
mds <- merge(data, pop_m, by.x="IID", by.y="id")
mds$col <- ifelse(mds$superpop=='AFR', 'darkolivegreen1', 
			ifelse(mds$superpop=='AMR', 'coral',
			ifelse(mds$superpop=='EAS', 'paleturquoise',
			ifelse(mds$superpop=='EUR', 'indianred1',
			ifelse(mds$superpop=='SAS', 'orchid1',
			ifelse(mds$superpop=='WHITE OR CAUCASIAN', 'firebrick2',
			ifelse(mds$superpop=='BLACK OR AFRICAN AMERICAN', 'darkgreen',
			ifelse(mds$superpop=='ASIAN', 'lightblue3',
			ifelse(mds$superpop=='hispanic', 'darkorange3', 'gray11')))))))))

col_ref <- unique(mds[,c('superpop','col')])
col_ref_tg <- col_ref[2:6,]
col_ref_als <- col_ref[1,]

pdf('MDS/MDS.pdf')
  mds_tg <- mds[mds$IID %in% als_pop$id,]
  mds_als <- mds[!mds$IID %in% als_pop$id,]
  #mds_als$superpop <- relevel(factor(mds_als$superpop), ref="WHITE OR CAUCASIAN")
  #mds_als <- mds_als[order(mds_als$superpop),]
  
  plot(mds_tg$C1, mds_tg$C2, col=mds_tg$col)
  points(mds_als$C1, mds_als$C2, col=mds_als$col, pch=15)
  legend('bottomright', title='1000 genomes groups', legend=col_ref_tg$superpop, fill=col_ref_tg$col)
  #legend('topright', title='ALS groups', legend=col_ref_als$superpop, fill=col_ref_als$col)
dev.off()


setwd('/nfs/turbo/bakulski1/People/johndou/ALS/')
write.csv(mds_als$IID, file='samples_passing_QC.csv', quote=F)

mds[mds$superpop!='WHITE OR CAUCASIAN' & grepl('FELD',mds$IID),]

keep <- mds_als[mds_als$IID=='FELD337' | mds_als$superpop=='WHITE OR CAUCASIAN',]
keep <- keep[!keep$V1 %in% c('FELD394','FELD494'),]
write.table(keep[,1:2], file='/nfs/turbo/bakulski1/People/johndou/ALS/keep_euro.txt', quote=F, row.names=F, col.names=F)
```