# Cumulative genetic risk and C9orf72 repeat status independently associate with ALS status in two case-control studies

John F. Dou1, Kelly M. Bakulski1, Kai Guo2,3, Junguk Hur4, Lili Zhou5, Sara Saez-Atienzar6, Ali R Stark6, Ruth Chia6, Alberto García-Redondo7,8, Ricardo Rojas-García8,9, Juan Francisco Vázquez-Costa8,10,11, Ruben Fernandez Santiago12,13,14, Sara Bandres-Ciga15, Pilar Gómez-Garre12,16, Maria Teresa Periñán12,16, Pablo Mir12,16,17, Jordi Pérez-Tur12,18,19, Fernando Cardona12,18,19, Manuel Menendez-Gonzalez20,21,22, Javier Riancho23,24, Daniel Borrego-Hernández7,8, Lucía Galán-Dávila25, Jon Infante Ceberio24, Pau Pastor26,27, Carmen Paradas28, Spanish Neurological Consortiuma, Bryan J. Traynor6, Eva L. Feldman2,3, and Stephen A Goutman2,3

Members of the Spanish Neurological Consortium: Jesús Esteban-Pérez7,8, Pilar Cordero-Vázquez7,8, Sevilla Teresa8,10,29, Adolfo López de Munain12,30,31, Julio Pardo-Fernández23,32, Ivonne Jericó-Pascual33,34, Oriol Dols-Icardo12,35 , Ellen Gelpi Mantius36,37, Janet Hoenicka8,38, Victoria Alvarez Martinez39,40, Francisco Javier Rodríguez de Rivera Garrido41, Katrin Beyer42, 43, Jordi Clarimón Echevarría12, 35



**Background**: Most amyotrophic lateral sclerosis (ALS) patients lack a monogenic mutation. This study evaluates ALS cumulative genetic risk in an independent Michigan and Spanish replication cohort using polygenic scores.
**Methods**: ALS (n=219) and healthy control (n=223) participant samples from University of Michigan were genotyped and assayed for the C9orf72 hexanucleotide expansion. Polygenic scores excluding the C9 region were generated using an independent ALS genome-wide association study (20,806 cases, 59,804 controls). Adjusted logistic regression and receiver operating characteristic curves evaluated the association and classification between polygenic scores and ALS status, respectively. Population attributable fractions and pathway analyses were conducted. An independent Spanish study sample (548 cases, 2,756 controls) was used for replication.
**Results**: Polygenic scores constructed from 275 single nucleotide polymorphisms had the best model fit in the Michigan cohort. A standard deviation increase in ALS polygenic score associated with 1.28 (95%CI 1.04-1.57) times higher odds of ALS with area under the curve of 0.663 versus a model without the ALS polygenic score (p-value=1x10-6). The population attributable fraction of the highest 20th percentile of ALS polygenic scores, relative to the lowest 80th percentile, was 4.1% of ALS cases. Genes annotated to this polygenic score enriched for important ALS pathomechanisms. Meta-analysis with the Spanish study, using a harmonized 132 single nucleotide polymorphism polygenic score, yielded similar logistic regression findings (odds ratio: 1.13, 95%CI 1.04-1.23).
**Conclusion**: ALS polygenic scores can account for cumulative genetic risk in populations and reflect disease-relevant pathways. If further validated, this polygenic score will inform future ALS risk models.
